<a href="https://colab.research.google.com/github/getolito/AlgoTech/blob/main/Yet_another_copy_of_Untitled58.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install openpyxl

In [ ]:
import pandas as pd
from scipy import stats

In [ ]:
import pandas as pd


     ID  Age  Revenu  Credit  Defaut
0     1   21     652   12635       0
1     2   43     654   14935       1
2     3   55    2117   14823       0
3     4   27    1330   14846       0
4     5   43    4010   17681       0
..  ...  ...     ...     ...     ...
95   96   60    2560    2623       0
96   97   37    4970    5260       1
97   98   68    2331    9401       0
98   99   52    4484   14832       1
99  100   68    1879   11640       0

[100 rows x 5 columns]


In [ ]:
# Affichage des 5 premieres lignes du dataset
df.head(5)

,ID,Age,Revenu,Credit,Defaut
0,1,21,652,12635,0
1,2,43,654,14935,1
2,3,55,2117,14823,0
3,4,27,1330,14846,0
4,5,43,4010,17681,0


In [ ]:
# Affichage du nombre d'observations (lignes)
print("Observations:", df.shape[0])

Observations: 100


In [ ]:
# Affichage du nombre de variables(colonnes)
nb_variables = df.shape[1]
print(f"Nombre de variables:{nb_variables}")

Nombre de variables:5


In [ ]:
# Identification des types de variables du dataset
print(df.dtypes)

ID        int64
Age       int64
Revenu    int64
Credit    int64
Defaut    int64
dtype: object


In [ ]:
import pandas as pd
df = pd.read_excel('base_100.xlsx')
# Calcul de l'age moyen des clients
col = 'Age'
print(f"Mean:{df[col].mean()}")

Mean:43.53


In [ ]:
#Calcul du revenu moyen des clients
col = 'Revenu'
print(f"Mean:{df[col].mean()}")

Mean:2671.61


In [ ]:
# Calcul du montant moyen du credit
col = 'Credit'
print(f"Mean:{df[col].mean()}")

Mean:11071.16


In [ ]:
# Identification du revenu minimum
col = 'Revenu'
print(f"Min:{df[col].min()}")

Min:556


In [ ]:
#Identification du revenu maximum
col = 'Revenu'
print(f"Max:{df[col].max()}")

Max:4985


In [ ]:
# Identification du credit minimum
col = 'Credit'
print(f"Min:{df[col].min()}")

Min:1159


In [ ]:
# Identification du credit maximum
col = 'Credit'
print(f"Max:{df[col].max()}")

Max:19971


In [ ]:
# Dispersion des revenus
col ='Revenu'
print(f"Range:{df[col].max()-df[col].min()}")

Range:4429


Nous constatons une forte variabilite des revenus, avec une difference de 4429 entre le revenu maximum et le revenu minimum.

In [ ]:
# dispersion des montants de credit
col = 'Credit'
print(f"Range:{df[col].max()-df[col].min()}")

Range:18812


Nous remarquons, apres calcul, une forte variabilite dans les montants de credit avec un ecart de 18812 entre le montant maximum de credit et le montant minimum de credit.

In [ ]:
# Comparaison du revenu moyen entre les deux groupes(Defaut = 0 vs Defaut = 1)
moyennes = df.groupby('Defaut')['Revenu'].mean()
print("Revenu moyen par groupe de defaut:")
print(moyennes)


Revenu moyen par groupe de defaut:
Defaut
0    2661.460526
1    2703.750000
Name: Revenu, dtype: float64


In [ ]:
# Comparaison du credit moyen entre les deux groupes(Defaut = 0 vs Defaut =1)
moyennes = df.groupby('Defaut')['Credit'].mean()
print("Credit moyen par groupe de defaut")
print(moyennes)

Credit moyen par groupe de defaut
Defaut
0    11026.407895
1    11212.875000
Name: Credit, dtype: float64


Structurons notre interpretation.Les clients qui sont en defaut de paiement c'est-a-dire qui ne remboursent pas leurs prets on les colle la valeur "Defaut" = 1 et ceux qui remboursent leurs prets, on les colle la valeur "Defaut = 1. Donc, Defaut = 0 (bon payeur), Defaut = 1 (mauvais payeur).
Apres examen, On constate que les revenus moyens des clients bon payeurs sont inferieurs aux revenus moyens des clients mauvais payeurs.

In [ ]:
import pandas as pd
from scipy import stats

In [ ]:
df = pd.read_excel('base_100.xlsx')
print(df)

     ID  Age  Revenu  Credit  Defaut
0     1   21     652   12635       0
1     2   43     654   14935       1
2     3   55    2117   14823       0
3     4   27    1330   14846       0
4     5   43    4010   17681       0
..  ...  ...     ...     ...     ...
95   96   60    2560    2623       0
96   97   37    4970    5260       1
97   98   68    2331    9401       0
98   99   52    4484   14832       1
99  100   68    1879   11640       0

[100 rows x 5 columns]


In [ ]:
# Creation de variables
df['Ratio_Credit_revenu'] = df['Credit']/df['Revenu'].replace(0,pd.NA)
print(df)

     ID  Age  Revenu  Credit  Defaut  Ratio_Credit_revenu
0     1   21     652   12635       0            19.378834
1     2   43     654   14935       1            22.836391
2     3   55    2117   14823       0             7.001889
3     4   27    1330   14846       0            11.162406
4     5   43    4010   17681       0             4.409227
..  ...  ...     ...     ...     ...                  ...
95   96   60    2560    2623       0             1.024609
96   97   37    4970    5260       1             1.058350
97   98   68    2331    9401       0             4.033033
98   99   52    4484   14832       1             3.307761
99  100   68    1879   11640       0             6.194784

[100 rows x 6 columns]


In [ ]:
# Calcul de la moyenne de ce Ratio
col = 'Ratio_Credit_revenu'
print(f"Mean:{df[col].mean()}")

Mean:6.018785685866012


In [ ]:
# Identification du client ayant le ratio le plus eleve
df['Ratio_Credit_revenu'] = df['Credit']/df['Revenu'].replace(0,pd.NA)
client_max_ratio=df.loc[df['Ratio_Credit_revenu'].idxmax()]
print("client ayant le ratio le plus eleve:")
print(client_max_ratio)

client ayant le ratio le plus eleve:
ID                        34.000000
Age                       22.000000
Revenu                   617.000000
Credit                 19971.000000
Defaut                     0.000000
Ratio_Credit_revenu       32.367909
Name: 33, dtype: float64


In [ ]:
import numpy as np

In [ ]:
# segmentation
# Definissons les seuils
seuil_bas = 2000
seuil_haut = 4000

In [ ]:
# Conditions
conditions = [
    df['Revenu']<seuil_bas,
    (df['Revenu']>=seuil_bas)&(df['Revenu']<seuil_haut),
    df['Revenu']>=seuil_haut
]

In [ ]:
# Choix
choix = ['Revenu faible', 'Revenu moyen','Revenu eleve']

In [ ]:
# Creation de la nouvelle variable
df['categorie_revenu'] = np.select(conditions,choix, default='')

In [ ]:
# affichage de la repartition par categorie
print(df['categorie_revenu'].value_counts())

categorie_revenu
Revenu moyen     43
Revenu faible    34
Revenu eleve     23
Name: count, dtype: int64


In [ ]:
# credit moyen par categorie de revenu
credit_moyen_par_categorie = df.groupby('categorie_revenu')['Credit'].mean()
print("Credit moyen par categorie:")
print(credit_moyen_par_categorie)

Credit moyen par categorie:
categorie_revenu
Revenu eleve     11293.869565
Revenu faible    11206.823529
Revenu moyen     10844.767442
Name: Credit, dtype: float64


In [ ]:
# taux de defaut par categorie
taux_defaut = df.groupby('categorie_revenu')['Defaut'].mean()*100
print("Taux de defaut par categorie de revenu(%):")
print(taux_defaut.round(2))

Taux de defaut par categorie de revenu(%):
categorie_revenu
Revenu eleve     26.09
Revenu faible    20.59
Revenu moyen     25.58
Name: Defaut, dtype: float64


In [ ]:
# extraction des clients ayant un credit superieur a 15000
clients_credit_eleve = df[df['Credit']>15000]
print("Clients avec un credit eleve:")
print(clients_credit_eleve)

Clients avec un credit eleve:
    ID  Age  Revenu  Credit  Defaut  Ratio_Credit_revenu categorie_revenu
4    5   43    4010   17681       0             4.409227     Revenu eleve
12  13   29    3770   17890       1             4.745358     Revenu moyen
13  14   47    2165   19695       0             9.096998     Revenu moyen
17  18   56    4742   19700       0             4.154365     Revenu eleve
19  20   47    1231   19328       0            15.701056    Revenu faible
20  21   33    3018   18405       0             6.098410     Revenu moyen
24  25   42    2958   15619       0             5.280257     Revenu moyen
31  32   32    4566   17201       1             3.767192     Revenu eleve
33  34   22     617   19971       0            32.367909    Revenu faible
37  38   53    4481   19014       0             4.243249     Revenu eleve
39  40   58     618   18509       0            29.949838    Revenu faible
40  41   18    1121   18552       1            16.549509    Revenu faible
49  50  

In [ ]:
# determination de leur age moyen
age_moyen_clients_eleves=clients_credit_eleve[['Age']]
print("Age moyen des clients avec un credit eleve:")
print(age_moyen_clients_eleves.mean())

Age moyen des clients avec un credit eleve:
Age    36.72
dtype: float64


In [ ]:
# determination de leur revenu moyen
revenu_moyen_clients_eleves=clients_credit_eleve[['Revenu']]
print("Revenu moyen des clients avec un credit eleve:")
print(revenu_moyen_clients_eleves.mean())

Revenu moyen des clients avec un credit eleve:
Revenu    2626.52
dtype: float64


In [ ]:
# Calcul de leur proportion de defaut
proportion_defaut_clients_eleves=clients_credit_eleve[['Defaut']]
print("Proportion de defaut des clients avec un credit eleve:")
print(proportion_defaut_clients_eleves.mean())

Proportion de defaut des clients avec un credit eleve:
Defaut    0.28
dtype: float64


In [ ]:
# Encodage
# creer une variable Risque, Faible si defaut = 0 , Eleve si Defaut = 1
df['Risque']=df['Defaut'].map({0:'Faible', 1:'Eleve'})
print(df[['Defaut','Risque']].head(10))

   Defaut  Risque
0       0  Faible
1       1   Eleve
2       0  Faible
3       0  Faible
4       0  Faible
5       1   Eleve
6       0  Faible
7       0  Faible
8       0  Faible
9       0  Faible


In [ ]:
# Verification de la transformation
print(df['Risque'].value_counts())

Risque
Faible    76
Eleve     24
Name: count, dtype: int64


In [ ]:
# Affichage de la frequence de chaque categorie
print(df['Risque'].value_counts(normalize=True*100))

Risque
Faible    0.76
Eleve     0.24
Name: proportion, dtype: float64


In [ ]:
# Analyse des extremes
# Identification des 3 clients ayant les credits les plus eleves
clients_credit_eleves=df.nlargest(3,'Credit')
print("Les 3 clients avec les credits les plus eleves:")
print(clients_credit_eleves)

Les 3 clients avec les credits les plus eleves:
    ID  Age  Revenu  Credit  Defaut  Ratio_Credit_revenu categorie_revenu  \
33  34   22     617   19971       0            32.367909    Revenu faible   
17  18   56    4742   19700       0             4.154365     Revenu eleve   
13  14   47    2165   19695       0             9.096998     Revenu moyen   

    Risque  
33  Faible  
17  Faible  
13  Faible  


In [ ]:
# Ayant les credits les plus faibles
clients_credit_eleves=df.nsmallest(3,'Credit')
print("Les 3 clients avec les credits les plus eleves:")
print(clients_credit_eleves)


Les 3 clients avec les credits les plus eleves:
    ID  Age  Revenu  Credit  Defaut  Ratio_Credit_revenu categorie_revenu  \
41  42   32    2243    1159       0             0.516719     Revenu moyen   
70  71   25    1920    1328       0             0.691667    Revenu faible   
55  56   28    4089    1618       0             0.395696     Revenu eleve   

    Risque  
41  Faible  
70  Faible  
55  Faible  


In [ ]:
# Coherence des donnees
# Existence de valeur manquantes
print(df.isnull().sum())

ID                     0
Age                    0
Revenu                 0
Credit                 0
Defaut                 0
Ratio_Credit_revenu    0
categorie_revenu       0
Risque                 0
dtype: int64


In [ ]:
# existence d'ages negatifs ou nuls
print(df[df['Age']<=0])



Empty DataFrame
Columns: [ID, Age, Revenu, Credit, Defaut, Ratio_Credit_revenu, categorie_revenu, Risque]
Index: []


In [ ]:
# existence de revenus negatifs
print(df[df['Revenu']<0])

Empty DataFrame
Columns: [ID, Age, Revenu, Credit, Defaut, Ratio_Credit_revenu, categorie_revenu, Risque]
Index: []


In [ ]:
# existence de credits negatis
print(df[df['Credit']<0])

Empty DataFrame
Columns: [ID, Age, Revenu, Credit, Defaut]
Index: []


Apres l'analyse des donnees, on constate qu'il n'existe aucune valeur manquante, pas d'ages negatifs ou nuls, pas de revenus et de credits negatifs non plus.

In [ ]:
# nombre de clients en defaut
nb_clients_defaut=df['Defaut'].sum()
print(f"Nombre de clients en defaut:{nb_clients_defaut}")


Nombre de clients en defaut:24


In [ ]:
# le nombre de clients sans defaut
nb_clients_sans_defaut=len(df)-nb_clients_defaut
print(f"Nombre de clients sans defaut:{nb_clients_sans_defaut}")

Nombre de clients sans defaut:76


In [ ]:
# taux global de defaut
taux_defaut_global=(nb_clients_defaut/len(df))*100
print(f"Taux global de defaut :{taux_defaut_global:.2f}")

Taux global de defaut :24.00


In [ ]:
import pandas as pd
# Rechargeons le jeux de donnees initial en vue d'eviter des erreurs dans les noms
df = pd.read_excel('base_100.xlsx')
print(df)
# Creons un nouveau fichie Excel contenant les donnees nettoyees avec les nouvelles variables : Ratio_Credit_Revenu, Categorie_Revenu,Risque
df.to_excel('dataset_nettoye.xlsx', index=False, sheet_name='base_100_analyse')
print("Apercu des donnees nettoyees")
print(df.head())

     ID  Age  Revenu  Credit  Defaut
0     1   21     652   12635       0
1     2   43     654   14935       1
2     3   55    2117   14823       0
3     4   27    1330   14846       0
4     5   43    4010   17681       0
..  ...  ...     ...     ...     ...
95   96   60    2560    2623       0
96   97   37    4970    5260       1
97   98   68    2331    9401       0
98   99   52    4484   14832       1
99  100   68    1879   11640       0

[100 rows x 5 columns]
Apercu des donnees nettoyees
   ID  Age  Revenu  Credit  Defaut
0   1   21     652   12635       0
1   2   43     654   14935       1
2   3   55    2117   14823       0
3   4   27    1330   14846       0
4   5   43    4010   17681       0


In [ ]:
# creation de Ratio_Credit_revenu
# evitons la division par zero
df['Ratio_Credit_revenu']=df['Credit']/df['Revenu'].replace(0,np.nan)
# creation de Categorie_Revenu
seuil_bas=2000 # en dessous:"Revenu faible"
seuil_haut=4000 # au-dessus : :"Revenu eleve"
# creation des intervalles
bins=[-float('inf'),seuil_bas,seuil_haut,float('inf')]
labels=['Revenu faible','Revenu moyen','Revenu eleve']
df['Categorie_Revenu']=pd.cut(df['Revenu'],bins=bins,labels=labels,right=False)
#creation de risque
df['Risque']=df['Defaut'].map({0:'Faible',1:'Eleve'})
# Rapide verification
print("Apercu des nouvelles colonnes")
print(df[['Credit','Revenu','Defaut','Ratio_Credit_revenu','Categorie_Revenu','Risque']].head(10))

Apercu des nouvelles colonnes
   Credit  Revenu  Defaut  Ratio_Credit_revenu Categorie_Revenu  Risque
0   12635     652       0            19.378834    Revenu faible  Faible
1   14935     654       1            22.836391    Revenu faible   Eleve
2   14823    2117       0             7.001889     Revenu moyen  Faible
3   14846    1330       0            11.162406    Revenu faible  Faible
4   17681    4010       0             4.409227     Revenu eleve  Faible
5   14911    1074       1            13.883613    Revenu faible   Eleve
6   12534    4103       0             3.054838     Revenu eleve  Faible
7    4817    3245       0             1.484438     Revenu moyen  Faible
8    6168    4375       0             1.409829     Revenu eleve  Faible
9   12347    2634       0             4.687547     Revenu moyen  Faible


In [ ]:
print("\nRepartition par categorie de revenu:")
print(df['Categorie_Revenu'].value_counts())
print("\nRepartition du risque:")
print(df['Risque'].value_counts())


Repartition par categorie de revenu:
Categorie_Revenu
Revenu moyen     43
Revenu faible    34
Revenu eleve     23
Name: count, dtype: int64

Repartition du risque:
Risque
Faible    76
Eleve     24
Name: count, dtype: int64


In [ ]:
# Jeux de donnees nettoyees avec les nouvelles variables
df.to_excel('dataset_nettoye.xlsx',index=False,sheet_name='base_100_analyse.xlsx')

In [ ]:
# Export du jeux de donnees enrichi
# Au format CSV
df.to_csv('dataset_avec_variables.csv',index=False,encoding='utf-8-sig')
print("\nFichier exporte:'dataset_avec_variables.csv")


Fichier exporte:'dataset_avec_variables.csv
